src/Symfony/Component/JsonStreamer/Read/PhpGenerator.php

```diff
@@ -109,20 +109,20 @@ private function generateProviders(DataModelNodeInterface $node, bool $decodeFro
 
             $arguments = $decodeFromStream ? '$stream, $offset, $length' : '$data';
 
-            $php .= $this->line("\$providers['".$node->getIdentifier()."'] = static function ($arguments) use (\$options, \$valueTransformers, \$instantiator, &\$providers) {", $context);
+            $php .= $this->line("\$providers[".$this->quote($node->getIdentifier())."] = static function ($arguments) use (\$options, \$valueTransformers, \$instantiator, &\$providers) {", $context);
 
             ++$context['indentation_level'];
 
             $php .= $decodeFromStream ? $this->line('$data = \\'.Decoder::class.'::decodeStream($stream, $offset, $length);', $context) : '';
 
             foreach ($node->getNodes() as $n) {
-                $value = $this->canBeDecodedWithJsonDecode($n, $decodeFromStream) ? $this->generateValueFormat($n, '$data') : '$providers[\''.$n->getIdentifier().'\']($data)';
+                $value = $this->canBeDecodedWithJsonDecode($n, $decodeFromStream) ? $this->generateValueFormat($n, '$data') : '$providers['.$this->quote($n->getIdentifier()).']($data)';
                 $php .= $this->line('if ('.$this->generateCompositeNodeItemCondition($n, '$data').') {', $context)
                     .$this->line("    return $value;", $context)
                     .$this->line('}', $context);
             }
 
-            $php .= $this->line('throw new \\'.UnexpectedValueException::class.'(\\sprintf(\'Unexpected "%s" value for "'.$node->getIdentifier().'".\', \\get_debug_type($data)));', $context);
+            $php .= $this->line('throw new \\'.UnexpectedValueException::class.'(\\sprintf(\'Unexpected "%s" value for "%s".\', \\get_debug_type($data), '.$this->quote($node->getIdentifier()).'));', $context);
```

Comment:
This is the modification that changes the generated code. You can also quote the full string so that the code is not modified and is slightly more optimized (even if optimizing error is pointless)


```suggestion
            $php .= $this->line('throw new \\'.UnexpectedValueException::class.'(\\sprintf('.$this->quote('Unexpected "%s" value for "'.$node->getIdentifier().'"').', \\get_debug_type($data)));', $context);
```

Reply:
Quoting an identifier in the same way a DBAL would adds the surrounding quotes to it (`$v = quote('id') === $v = "'id'"`). By doing so, the identifier becomes a valid piece of code to "inject" into a statement rather than just a plain string to interpolate. In this specific case, what’s being done is an interpolation. Trying to merge both worlds would come at the cost of readability. Moreover, since PHP 8.4, `sprintf` has been optimized to inline those cases automatically. But, I prefer to prioritize the readability of the code generation logic (which can already be quite complex to follow) over optimizing the generation of the exception message that is thrown only once at runtime.